ml-proto/src/spec/arithmetic.ml

```diff
@@ -119,10 +123,36 @@ struct
 
   let unop op =
     let f = match op with
-      | Clz -> fun i -> i  (* TODO *)
-      | Ctz -> fun i -> i  (* TODO *)
-      | Popcnt -> fun i -> i  (* TODO *)
-    in fun v -> Int.to_value (f (Int.of_value 1 v))
+      | Clz -> fun (v : Int.t) -> 
+                 let rec helper n acc =
+                   match n with 
+                     | n when n = Int.zero -> Int.size 
+                     | n when Int.logand n (Int.shift_left Int.one (Int.size-1))
+                              <> Int.zero
+                         -> acc 
+                     | _ -> helper (Int.shift_left n 1) (1+acc)
+                 in helper v 0
+      | Ctz -> fun v ->
+                 let rec helper n acc =
+                   match n with 
+                     | n when n = Int.zero -> Int.size
+                     | n when Int.logand n Int.one = Int.one -> acc
+                     | _ -> helper (Int.shift_right n 1) (1+acc)
+                 in helper v 0
+      | Popcnt -> fun v ->
+                    let rec helper n acc i =
+                      match (n, i) with 
+                       | (n,_) when n = Int.zero -> acc
+                       | (_,0) -> acc
```

Comment:
The first line matches n=Int.zero and the second when i=0. It would have been clearer and more succinct if this could be have turned into an OR pattern but alas OR patterns are incompatible with guards. OCaml will helpfully identify some redundant or non-exhaustive pattern cases.